Use project asset for project image uploads

tnagorra opened the pull request:
## Changes

- Project Type
  - Add support for completeness project
  - Fix issue with project query with project_type_specifics field
- Project Asset
  - Use ProjectAsset as image
  - Remove image from ProjectCreate serializer
  - Update image to receive ProjectAsset on ProjectUpdate and ProcessedProjectUpdate 
serializer
  - Add validation for image on both serializers
  - Append ULID suffix on uploaded asset filename
- Project Processing
  - Cleanup tasks and groups when transitioning from MARKED_AS_READY
  - Flag all assets as to be deleted (except INPUT types)


## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests

tnagorra committed:
Use project asset for project image uploads

tnagorra committed:
Add a is_deleted field on ProjectAsset

- Filter out is_deleted fields using ModelManager
- Cleanup project asset, groups and tasks when processing a project

tnagorra committed:
Use ulid on uploaded file names

- Add extention on temporary image file on test

tnagorra committed:
Add support for completeness project

tnagorra committed:
Fix issue with project_type_specifics query

- Update test for project query
- Move to_camel_case and to_snake_case functions to utils.common

tnagorra committed:
Add tile_x, tile_y, tile_z information on the intermediate export

susilnem commented on apps/project/graphql/types.py:
```diff
@@ -1,13 +1,41 @@
 from utils.geo.tile_server.models import TileServerCommonConfig, TileServerConfig, TileServerCustomConfig
 
 
+def prepare_tile_server_property(prop: dict | None):
```
```suggestion
def prepare_tile_server_property(prop: dict | None) -> ProjectTileServerConfig | None:
```

tnagorra committed:
Use pydantic class for project_type_specifics query

- Instead of strawberry serializer class, use pydantic class for
  initialization.

thenav56 commented on apps/project/models.py:
```diff
@@ -327,6 +343,11 @@ class ProjectAsset(UserResource):
         related_name="+",
     )
 
+    is_deleted = models.BooleanField(
+        default=False,
+        help_text=gettext_lazy("If this flag is enabled, this project asset will be deleted in the future"),
+    )
```
```suggestion
    marked_as_delete = models.BooleanField(
        default=False,
        help_text=gettext_lazy("If this flag is enabled, this project asset will be deleted in the future"),
    )
```

thenav56 commented on apps/project/models.py:
```diff
@@ -304,10 +311,19 @@ def clean(self):
         #     self.requiredResults += group.requiredCount * group.numberOfTasks
 
 
+class SoftDeletionManager(models.Manager):
+    @typing.override
+    def get_queryset(self):
+        return super().get_queryset().filter(is_deleted=False)
```
Let's not use this.... client should handle this

Add a validation in the project serializer... (Or unset is_deleted if used)

thenav56 commented on apps/project/project_types/tile_map_service/base/project.py:
```diff
@@ -43,19 +43,21 @@ class TileMapServiceProjectProperty(base_project.BaseProjectProperty):
         project_id = info.context.get("project_id")
-        # FIXME(tnagorra): Handle error
-        if not ProjectAsset.objects.filter(
+        asset_exists = ProjectAsset.objects.filter(
```
Use a method for non-deleted asset....

thenav56 commented on apps/project/serializers.py:
```diff
@@ -88,6 +96,24 @@ def validate_status(self, new_status: Project.Status):
+    def validate_image(self, new_image: ProjectAsset):
+        assert self.instance is not None
+
+        asset_exists = ProjectAsset.objects.filter(
```
Use custom method for non-deleted qs

thenav56 commented on apps/project/serializers.py:
```diff
@@ -182,6 +210,24 @@ def validate_status(self, new_status: Project.Status):
+    def validate_image(self, new_image: ProjectAsset):
+        assert self.instance is not None
+
+        asset_exists = ProjectAsset.objects.filter(
```
Use custom method for non-deleted qs

thenav56 commented on apps/project/serializers.py:
```diff
@@ -194,3 +240,11 @@ class Meta:  # type: ignore[reportIncompatibleVariableOverride]
             "project",
         )
+
+    def validate_type(self, new_type: Project.Status):
```
Let's move this to def create(..) and add static type there

thenav56 commented on apps/project/tests/mutation_test.py:
```diff
@@ -23,61 +23,76 @@
         )
 
 
+def create_project_query(
+    *,
+    query_check_func: typing.Callable,
+    query: str,
+    project_data: dict,
+    **kwargs,
+) -> dict:
+    return query_check_func(
+        query,
+        variables={
+            "data": project_data,
+        },
+        **kwargs,
+    )
```
Let's remove this
